src/Agent.Worker/Container/ContainerInfo.cs

```diff
@@ -144,11 +207,124 @@ public MountVolume(string sourceVolumePath, string targetVolumePath, bool readOn
+            else if (volumeSplit.Length == 2)
+            {
+                if (String.Equals(volumeSplit[1], "ro", StringComparison.OrdinalIgnoreCase))
+                {
+                    // target:ro
+                    TargetVolumePath = volumeSplit[0];
+                    ReadOnly = true;
+                }
+                else
+                {
+                    // source:target
+                    SourceVolumePath = volumeSplit[0];
+                    TargetVolumePath = volumeSplit[1];
+                    ReadOnly = false;
+                }
+            }
+            else
+            {
+                // target - or, default to passing straight through
+                TargetVolumePath = volume;
+                ReadOnly = false;
+            }
         }
 
         public string SourceVolumePath { get; set; }
         public string TargetVolumePath { get; set; }
         public bool ReadOnly { get; set; }
+        public bool IsExpanded { get; set; }
+        public string Raw { get; set; }
+    }
+
+    public class PortMapping
+    {
+        public PortMapping(string hostPort, string containerPort, string protocol)
+        {
+            this.HostPort = hostPort;
+            this.ContainerPort = containerPort;
+            this.Protocol = protocol;
+            this.IsExpanded = true;
+        }
+
+        public PortMapping(string raw, bool isExpanded)
+        {
+            this.IsExpanded = isExpanded;
+            if (this.IsExpanded)
+            {
+                ParsePortString(raw);
+            }
+            else
+            {
+                this.Raw = raw;
+            }
+        }
+
+        private void ParsePortString(string port)
+        {
+            var protoSplit = port.Split("/");
+            String portString;
+            String protoString = null;
+            if (protoSplit.Length == 2)
+            {
+                protoString = protoSplit[1];
+            }
+            portString = protoSplit[0];
+            var portSplit = portString.Split(":");
+            if (portSplit.Length == 3)
+            {
+                // host:hostport:targetport
+                HostPort = $"{portSplit[0]}:{portSplit[1]}";
+                ContainerPort = portSplit[2];
+                Protocol = protoString;
+            }
+            else if (portSplit.Length == 2)
+            {
+                // hostport:targetport
+                HostPort = portSplit[0];
+                ContainerPort = portSplit[1];
+                Protocol = protoString;
+            }
+            else
+            {
+                // target - or, default to passing straight through
+                ContainerPort = port;
+                Protocol = protoString;
+            }
+        }
+
+        public string HostPort { get; set; }
+        public string ContainerPort { get; set; }
+        public string Protocol { get; set; }
+        public bool IsExpanded { get; set; }
+        public string Raw { get; set; }
```

Comment:
is `IsExpanded` and `Raw` used publicly?

Reply:
I use them in the JobRunner for the delayed expansion of variables. I didnt feel as good about moving all the variable expansion logic into where they could remain private

I think a better solution in the future might be to overload the variable expansion to take more than just a dictionary of variables, so I can pass in the list of mappings and have them expanded in place instead of building the dictionary to pass.

Reply:
No longer used